okta/okta.go

```diff
@@ -0,0 +1,242 @@
+}
+
+func (m *oktaUserManager) CreateUserToken(user string, password string) (string, error) {
+	return "", errors.New("not implemented")
+}
+
+func (m *oktaUserManager) GetLoginHandler(callbackURL string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO (kim): persist nonce and state.
+		// nonce, err := generateNonce()
+		// if err != nil {
+		//     grip.Error(message.WrapError(err, message.Fields{
+		//         "message": "could not get login handler",
+		//     }))
+		//     gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(errors.Wrap(err, "could not get login handler")))
+		//     return
+		// }
+		q := r.URL.Query()
+		q.Add("client_id", m.clientID)
+		q.Add("response_type", "code")
+		q.Add("response_mode", "query")
+		q.Add("scope", "openid")
+		q.Add("redirect_uri", m.redirectURI)
+		q.Add("state", "TODO")
+		q.Add("nonce", "TODO")
+
+		http.Redirect(w, r, fmt.Sprintf("%s/oauth2/v1/authorize?%s", m.issuer, q.Encode()), http.StatusMovedPermanently)
+	}
+}
+
+func (m *oktaUserManager) GetLoginCallbackHandler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// TODO (kim): verify state.
+		// checkState := r.URL.Query().Get("state")
+		// state := "TODO"
+		// if state != checkState {
+		//     grip.Error(message.WrapError(err, message.Fields{
+		//         "message":        "mismatched states during authentication",
+		//         "context":        "Okta",
+		//         "expected_state": state,
+		//         "actual_state":   checkState,
+		//     }))
+		//     gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(errors.New("mismatched states during authentication")))
+		//     return
+		// }
+
+		if errCode := r.URL.Query().Get("error"); errCode != "" {
+			desc := r.URL.Query().Get("error_description")
+			err := fmt.Errorf("%s: %s", errCode, desc)
+			grip.Error(message.WrapError(errors.WithStack(err), message.Fields{
+				"message": "failure in callback handler redirect",
+				"op":      "GetLoginCallbackHandler",
+				"auth":    "Okta",
+			}))
+			gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(errors.Wrap(err, "could not get login callback handler")))
+			return
+		}
+
+		resp, err := m.getToken(r.URL.Query().Get("code"))
+		if err != nil {
+			err = errors.Wrap(err, "could not get ID token")
+			gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(err))
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "failed to get token from Okta",
+				"op":      "GetLoginCallbackHandler",
+				"auth":    "Okta",
+			}))
+			return
+		}
+		if err := m.validateToken(resp.IDToken); err != nil {
+			err = errors.Wrap(err, "could not validate ID token from Okta")
+			gimlet.WriteResponse(w, gimlet.MakeTextErrorResponder(err))
+			grip.Error(message.WrapError(err, message.Fields{
+				"message": "failed to validate ID token",
+				"op":      "GetLoginCallbackHandler",
+				"auth":    "Okta",
+			}))
+		}
+		// TODO (kim): persist token.
```

Comment:
Feel free to kick this to later, but out of curiosity what does this token represent? Where are you planning on persisting it?